### PR TITLE
BUG: fix calling local references with keyword arguments in query

### PR DESCRIPTION
**Pull request by danielhrisca**
- [ ] closes #xxxx
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

calling local references with keyword arguments in query fails with attribute error:

```python
import numpy as np
import pandas as pd

def f(a, b):
    return b

df = pd.DataFrame()

df['val'] = np.arange(10)
print(df.query('val > @f(4, 5)'))
print(df.query('val > @f(4, b=5)'))
```

```
   val
6    6
7    7
8    8
9    9
Traceback (most recent call last):

  File "C:/Users/uidn3651/Desktop/untitled0.py", line 10, in <module>
    print(df.query('val > @f(4, b=5)'))

  File "d:\work\02__pythonworkspace\_venv\lib\site-packages\pandas\core\frame.py", line 3052, in query
    res = self.eval(expr, **kwargs)

  File "d:\work\02__pythonworkspace\_venv\lib\site-packages\pandas\core\frame.py", line 3169, in eval
    return _eval(expr, inplace=inplace, **kwargs)

  File "d:\work\02__pythonworkspace\_venv\lib\site-packages\pandas\core\computation\eval.py", line 293, in eval
    truediv=truediv)

  File "d:\work\02__pythonworkspace\_venv\lib\site-packages\pandas\core\computation\expr.py", line 749, in __init__
    self.terms = self.parse()

  File "d:\work\02__pythonworkspace\_venv\lib\site-packages\pandas\core\computation\expr.py", line 766, in parse
    return self._visitor.visit(self.expr)

  File "d:\work\02__pythonworkspace\_venv\lib\site-packages\pandas\core\computation\expr.py", line 370, in visit
    return visitor(node, **kwargs)

  File "d:\work\02__pythonworkspace\_venv\lib\site-packages\pandas\core\computation\expr.py", line 376, in visit_Module
    return self.visit(expr, **kwargs)

  File "d:\work\02__pythonworkspace\_venv\lib\site-packages\pandas\core\computation\expr.py", line 370, in visit
    return visitor(node, **kwargs)

  File "d:\work\02__pythonworkspace\_venv\lib\site-packages\pandas\core\computation\expr.py", line 379, in visit_Expr
    return self.visit(node.value, **kwargs)

  File "d:\work\02__pythonworkspace\_venv\lib\site-packages\pandas\core\computation\expr.py", line 370, in visit
    return visitor(node, **kwargs)

  File "d:\work\02__pythonworkspace\_venv\lib\site-packages\pandas\core\computation\expr.py", line 674, in visit_Compare
    return self.visit(binop)

  File "d:\work\02__pythonworkspace\_venv\lib\site-packages\pandas\core\computation\expr.py", line 370, in visit
    return visitor(node, **kwargs)

  File "d:\work\02__pythonworkspace\_venv\lib\site-packages\pandas\core\computation\expr.py", line 476, in visit_BinOp
    op, op_class, left, right = self._maybe_transform_eq_ne(node)

  File "d:\work\02__pythonworkspace\_venv\lib\site-packages\pandas\core\computation\expr.py", line 413, in _maybe_transform_eq_ne
    right = self.visit(node.right, side='right')

  File "d:\work\02__pythonworkspace\_venv\lib\site-packages\pandas\core\computation\expr.py", line 370, in visit
    return visitor(node, **kwargs)

  File "d:\work\02__pythonworkspace\_venv\lib\site-packages\pandas\core\computation\expr.py", line 657, in visit_Call
    kwargs.append(ast.keyword(

AttributeError: 'dict' object has no attribute 'append'
```

With the fix this works correctly

```
   val
6    6
7    7
8    8
9    9
   val
6    6
7    7
8    8
9    9
```

### CODE REVIEW

**Review by jreback — CHANGES_REQUESTED** [review]
tests always are first

**Comment by danielhrisca** [comment]
@jreback 

running _test_fast.bat_ produces many failed tests. However I can't find _test_multi_line_expression_callable_local_variable_ in the output. Does this mean that this test is passed?

**Comment by danielhrisca** [comment]
@jreback 
can you have a look at the new commit? Thank you

File: pandas/tests/computation/test_eval.py [hunk 1]
```diff
@@ -1345,6 +1345,33 @@ def test_multi_line_expression_local_variable(self):
...
+        expected = df.copy()
+        expected['c'] = expected['a'] * local_func(b=7, a=1)
+        expected['d'] = expected['c'] + local_func(b=7, a=1)
+        ans = df.eval("""
+        c = a * @local_func(b=7, a=1)
+        d = c + @local_func(b=7, a=1)
+        """, inplace=True)
+        assert_frame_equal(expected, df)
+        assert ans is None
+
```
Comment (gfyoung) [inline pandas/tests/computation/test_eval.py]: Split this test into two.

Comment (danielhrisca) [reply pandas/tests/computation/test_eval.py]: @gfyoung 
done

File: pandas/tests/computation/test_eval.py [hunk 2]
```diff
@@ -1345,6 +1345,38 @@ def test_multi_line_expression_local_variable(self):
         assert_frame_equal(expected, df)
         assert ans is None
 
+    def test_multi_line_expression_callable_local_variable(self):
```
Comment (jreback) [inline pandas/tests/computation/test_eval.py]: so this works currently yes?

File: pandas/tests/computation/test_eval.py [hunk 3]
```diff
@@ -1345,6 +1345,38 @@ def test_multi_line_expression_local_variable(self):
...
+        expected['d'] = expected['c'] + local_func(1, 7)
+        ans = df.eval("""
+        c = a * @local_func(1, 7)
+        d = c + @local_func(1, 7)
+        """, inplace=True)
+        assert_frame_equal(expected, df)
+        assert ans is None
+
+    def test_multi_line_expression_callable_local_variable_with_kwargs(self):
+        df = pd.DataFrame({'a': [1, 2, 3], 'b': [4, 5, 6]})
```
Comment (jreback) [inline pandas/tests/computation/test_eval.py]: add the issue reference (is there an issue open for this? if not add the pr number) as a comment

**Review by jreback — CHANGES_REQUESTED** [review]
add a whatsnew note; reshaping or other section would be ok

File: pandas/tests/computation/test_eval.py (hunk 2 above)
Comment (danielhrisca) [reply pandas/tests/computation/test_eval.py]: Prior to this fix only positional arguments were working.

File: pandas/tests/computation/test_eval.py (hunk 3 above)
Comment (danielhrisca) [reply pandas/tests/computation/test_eval.py]: I'm not sure what I'm supposed to add here. Something like ``# Github PR #26426 `` ?

I couldn't find a related issue.

File: pandas/tests/computation/test_eval.py (hunk 2 above)
Comment (danielhrisca) [reply pandas/tests/computation/test_eval.py]: > add a whatsnew note; reshaping or other section would be ok

I've added an entry in the _other_ section

File: doc/source/whatsnew/v0.25.0.rst [hunk 4]
```diff
@@ -475,6 +474,7 @@ Other
 - Removed unused C functions from vendored UltraJSON implementation (:issue:`26198`)
 - Bug in :func:`factorize` when passing an ``ExtensionArray`` with a custom ``na_sentinel`` (:issue:`25696`).
 - Allow :class:`Index` and :class:`RangeIndex` to be passed to numpy ``min`` and ``max`` functions.
+- Allow keyword arguments for callable local reference used in the ``DataFrame.query`` string
```
Comment (jreback) [inline doc/source/whatsnew/v0.25.0.rst]: use ``:method:`DataFrame.query` ``; add the issue number (I guess no issue, so this PR number);

move this to indexing section

**Review by jreback — CHANGES_REQUESTED** [review]

File: doc/source/whatsnew/v0.25.0.rst (hunk 4 above)
Comment (danielhrisca) [reply doc/source/whatsnew/v0.25.0.rst]: I've moved the entry in the indexing section

File: pandas/tests/computation/test_eval.py (hunk 3 above)
Comment (jreback) [reply pandas/tests/computation/test_eval.py]:
```
# 26526
```

is enough

**Comment by jreback** [comment]
you will need to merge master as well, there is a conflict

**Comment by danielhrisca** [comment]
Th conflict is now resolved

File: doc/source/whatsnew/v0.25.0.rst [hunk 5]
```diff
@@ -380,6 +379,7 @@ Indexing
 - Bug in which :meth:`DataFrame.append` produced an erroneous warning indicating that a ``KeyError`` will be thrown in the future when the data to be appended contains new columns (:issue:`22252`).
 - Bug in which :meth:`DataFrame.to_csv` caused a segfault for a reindexed data frame, when the indices were single-level :class:`MultiIndex` (:issue:`26303`).
 - Fixed bug where assigning a :class:`arrays.PandasArray` to a :class:`pandas.core.frame.DataFrame` would raise error (:issue:`26390`)
+- Allow keyword arguments for callable local reference used in the :method:`DataFrame.query` string (PR 26426)
```
Comment (jreback) [inline doc/source/whatsnew/v0.25.0.rst]: this needs to be like the other, e.g. ``(:issue:`26426`)``

**Review by jreback — CHANGES_REQUESTED** [review]
need a different format for the issue number

File: doc/source/whatsnew/v0.25.0.rst (hunk 5 above)
Comment (danielhrisca) [reply doc/source/whatsnew/v0.25.0.rst]: Done. Thank you for the patience!

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks for the patch @danielhrisca

**Pull request merged**